File: Editorial/0120_Triangle/CPP/2_Dynamic_Programming.cpp
class Solution {
public:
    int minimumTotal(vector<vector<int>>& triangle) {
        vector<int> prevRow = triangle[0];
        for (int row = 1; row < triangle.size(); row++) {
            vector<int> currRow;
            for (int col = 0; col <= row; col++) {
                int smallestAbove = INT_MAX;
                if (col > 0) {
                    smallestAbove = prevRow[col - 1];
                }
                if (col < row) {
                    smallestAbove = min(smallestAbove, prevRow[col]);
                }
                currRow.push_back(smallestAbove + triangle[row][col]);
            }
            prevRow = currRow;
        }
        return *min_element(prevRow.begin(), prevRow.end());
    }
};

File: Editorial/0101-0200/0120_Triangle/CSharp/2_Dynamic_Programming.cs
public class Solution {
    public int MinimumTotal(IList<IList<int>> triangle) {
        List<int> prevRow = triangle[0].ToList();
        for (int row = 1; row < triangle.Count; row++) {
            List<int> currRow = new List<int>();
            for (int col = 0; col <= row; col++) {
                int smallestAbove = int.MaxValue;
                if (col > 0) {
                    smallestAbove = prevRow[col - 1];
                }

                if (col < row) {
                    smallestAbove = Math.Min(smallestAbove, prevRow[col]);
                }

                currRow.Add(smallestAbove + triangle[row][col]);
            }

            prevRow = currRow;
        }

        return prevRow.Min();
    }
}
